### DISCUSSION TITLE
Convert Laurent polynomial to polynomial?

### DISCUSSION THREAD

**Discussion by oskarhenriksson** [oscar-system/Oscar.jl#3156] [Q&A] [answered]
Suppose that I have a Laurent polynomial $f\in k[x^\pm]$ where all exponents happen to be nonnegative, say

```
R, t = LaurentPolynomialRing(QQ, "t")
f = 5*t^3 - 7*t + 1
````

and that I want to view it as a polynomial $g\in k[x]$. Is there a nice way to do this in Oscar?

Naively, I would have hoped to be able to form the inclusion $\varphi\colon k[x]\hookrightarrow k[x^\pm]$ and use the `preimage` command, like this:

```
S, t = polynomial_ring(QQ,["t"])
phi = hom(S,R,[t])
g = preimage(phi, f)
````

but `R` doesn't seem to have the correct datatype for it to be possible to set up algebra homomorphisms like this.

(_Sidenote:_ Part of the reason I want to do this is because I want to be able to find the roots of $f$ with the `roots` command. If there is a smooth way to do this without passing via $k[t]$, that would also be of interest!)

**Comment by fieker** [comment, marked as answer]
On Mon, Jan 08, 2024 at 04:21:34AM -0800, Oskar Henriksson wrote:
> Suppose that I have a Laurent polynomial $f\in k[x^\pm]=R$ where all exponents happen to be nonnegative, say
> 
> ```
> R, t = LaurentPolynomialRing(QQ, "t")
> f = 5*t^3 - 7*t + 1
> ````
> 
> and that I want to view it as a polynomial in the subring of regular polynomials, $S=k[x]$. Is there a nice way to do this in Oscar?
> 
> Naively, I would have hoped that this would work:
> 
> ```
> S, t = polynomial_ring(QQ,["t"])
> phi = hom(S,R,[t])
> g = preimage(phi, f)
> ````
> 
> but `R` doesn't seem to have the correct datatype for it to be possible to set up algebra homomorphisms like this.
> 
> (_Sidenote:_ Part of the reason I want to do this is because I want to be able to find the roots of $f$ with the `roots` command. If there is a smooth way to do this without passing via $k[t]$, that would also be of interest!)

In your case, for a quick answer:
f.poly
is what you want - provided f.mindeg is, as here, 0.
We need to provide a better interface for that

**Comment by thofma** [comment]
1. I agree, that we need to improve the univariate experience and bring it on par with the multivariate one.
2. Maybe it makes sense to extend the `roots` function to Laurent polynomials. Then the only question is what to do about zero. Always exclude it? E.g. should we have
   ```
   R, t = LaurentPolynomialRing(QQ, "t")
   roots(t) == [] # or
   roots(t) == [zero(R)]
   ```
